
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>codonvarianttable &#8212; dms_variants 0.4.5 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="constants" href="dms_variants.constants.html" />
    <link rel="prev" title="binarymap" href="dms_variants.binarymap.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-dms_variants.codonvarianttable"></span><div class="section" id="codonvarianttable">
<h1>codonvarianttable<a class="headerlink" href="#codonvarianttable" title="Permalink to this headline">¶</a></h1>
<p>Defines <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> objects for storing and
handling codon variants of a gene.</p>
<dl class="class">
<dt id="dms_variants.codonvarianttable.CodonVariantTable">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.codonvarianttable.</code><code class="sig-name descname">CodonVariantTable</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">barcode_variant_file</em>, <em class="sig-param">geneseq</em>, <em class="sig-param">substitutions_are_codon=False</em>, <em class="sig-param">extra_cols=None</em>, <em class="sig-param">substitutions_col='substitutions'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Store and handle counts of barcoded codon variants of a gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>barcode_variant_file</strong> (<em>str</em>) – CSV file giving barcodes and variants. Must have columns “library”,
“barcode”, “substitutions” (nucleotide mutations in 1, …, numbering
in a format like “G301A A302T G856C”), and “variant_call_support”
(sequences supporting barcode-variant call). Additional columns are
removed unless they are specified <cite>extra_cols</cite>.</p></li>
<li><p><strong>geneseq</strong> (<em>str</em>) – Sequence of wildtype protein-coding gene.</p></li>
<li><p><strong>substitutions_are_codon</strong> (<em>bool</em>) – If <cite>True</cite>, then “substitutions” column in <cite>barcode_variant_file</cite> gives
substitutions as codon rather than nucleotide mutations (e.g.,
“ATG1ATA GTA5CCC” for substitutions at codons 1 and 5.</p></li>
<li><p><strong>extra_cols</strong> (<em>list</em>) – Additional columns in <cite>barcode_variant_file</cite> to retain when creating
<cite>barcode_variant_df</cite> and <cite>variant_count_df</cite> attributes.</p></li>
<li><p><strong>substitutions_col</strong> (<em>str</em>) – Name of substitutions column in <cite>barcode_variant_file</cite> (use if you
want it to be something other than “substitutions”).</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.geneseq">
<code class="sig-name descname">geneseq</code><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.geneseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Wild-type sequence passed at initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.sites">
<code class="sig-name descname">sites</code><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.sites" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all codon sites in 1, 2, … numbering.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.codons">
<code class="sig-name descname">codons</code><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.codons" title="Permalink to this definition">¶</a></dt>
<dd><p><cite>codons[r]</cite> is wildtype codon at site <cite>r</cite>, ordered by sites.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>collections.OrderedDict</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.aas">
<code class="sig-name descname">aas</code><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.aas" title="Permalink to this definition">¶</a></dt>
<dd><p><cite>aas[r]</cite> is wildtype amino acid at site <cite>r</cite>, ordered by sites.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>collections.OrderedDict</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.libraries">
<code class="sig-name descname">libraries</code><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.libraries" title="Permalink to this definition">¶</a></dt>
<dd><p>List of libraries in <cite>barcode_variant_file</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.barcode_variant_df">
<code class="sig-name descname">barcode_variant_df</code><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.barcode_variant_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Info about codon mutations parsed from <cite>barcode_variant_file</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.variant_count_df">
<code class="sig-name descname">variant_count_df</code><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.variant_count_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Initially <cite>None</cite>, but after data added with
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts" title="dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.addSampleCounts</span></code></a>, holds counts of
each variant for each sample. Differs from <cite>barcode_variant_df</cite>
in that the former just holds barcode-variant definitions,
whereas <cite>variant_count_df</cite> has counts for each sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pandas.DataFrame or None</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.addMergedLibraries">
<em class="property">static </em><code class="sig-name descname">addMergedLibraries</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">*</em>, <em class="sig-param">all_lib='all libraries'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.addMergedLibraries"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.addMergedLibraries" title="Permalink to this definition">¶</a></dt>
<dd><p>Add data to <cite>df</cite> for all libraries merged.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data frame that includes columns named ‘library’ and ‘barcode’.</p></li>
<li><p><strong>all_lib</strong> (<em>str</em>) – Name given to library that is merge of all other libraries.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If <cite>df</cite> only has data for one library, just returns <cite>df</cite>. Otherwise
returns copy of <cite>df</cite> with new library with name given by <cite>all_lib</cite>
that contains data for all individual libraries with the ‘barcode’
column giving library name followed by a hyphen and barcode.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts">
<code class="sig-name descname">addSampleCounts</code><span class="sig-paren">(</span><em class="sig-param">library</em>, <em class="sig-param">sample</em>, <em class="sig-param">barcodecounts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.addSampleCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Add variant counts for a sample to <cite>variant_count_df</cite>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you have many samples to add at once, it is faster to use
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df" title="dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.add_sample_counts_df()</span></code></a> rather than
repeatedly calling this method.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>library</strong> (<em>str</em>) – Valid <cite>library</cite> for the <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>.</p></li>
<li><p><strong>sample</strong> (<em>str</em>) – Sample name, must <strong>not</strong> already be in
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.samples" title="dms_variants.codonvarianttable.CodonVariantTable.samples"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.samples</span></code></a> for <cite>library</cite>.</p></li>
<li><p><strong>barcodecounts</strong> (<em>pandas.DataFrame</em>) – Counts for each variant by barcode. Must have columns “barcode”
and “count”. The “barcode” column must have all barcodes in
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.valid_barcodes" title="dms_variants.codonvarianttable.CodonVariantTable.valid_barcodes"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.valid_barcodes</span></code></a> for <cite>library</cite>. Such
data frames are returned by
<a class="reference internal" href="dms_variants.illuminabarcodeparser.html#dms_variants.illuminabarcodeparser.IlluminaBarcodeParser" title="dms_variants.illuminabarcodeparser.IlluminaBarcodeParser"><code class="xref py py-mod docutils literal notranslate"><span class="pre">dms_variants.illuminabarcodeparser.IlluminaBarcodeParser</span></code></a>.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.add_full_seqs">
<code class="sig-name descname">add_full_seqs</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">*</em>, <em class="sig-param">aa_seq_col='aa_sequence'</em>, <em class="sig-param">codon_seq_col='codon_sequence'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.add_full_seqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.add_full_seqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Add full sequences to data frame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Data frame that specifies substitutions. Might be the
<cite>barcode_variant_df</cite> or <cite>variant_count_df</cite> attributes,
or the result of calling <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.func_scores" title="dms_variants.codonvarianttable.CodonVariantTable.func_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.func_scores()</span></code></a>.</p></li>
<li><p><strong>aa_seq_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with amino-acid sequences. In order to add,
<cite>df</cite> must have column named ‘aa_substitutions’.</p></li>
<li><p><strong>codon_seq_col</strong> (<em>str</em><em> or </em><em>None</em>) – Name of added column with codon sequences. In order to add,
<cite>df</cite> must have column named ‘codon_substitutions’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Copy of <cite>df</cite> with columns <cite>aa_seq_col</cite> and/or <cite>codon_seq_col</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df">
<code class="sig-name descname">add_sample_counts_df</code><span class="sig-paren">(</span><em class="sig-param">counts_df</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.add_sample_counts_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.add_sample_counts_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Add variant counts for several samples to <cite>variant_count_df</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>counts_df</strong> (<em>pandas.DataFrame</em>) – Must have columns ‘library’, ‘sample’, ‘barcode’, and ‘count’.
The sample must <strong>not</strong> already be in <cite>CodonVariantTable.samples</cite>
for that library. The barcode columns must have all barcodes for
that library including zero-count ones.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.avgCountsPerVariant">
<code class="sig-name descname">avgCountsPerVariant</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.avgCountsPerVariant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.avgCountsPerVariant" title="Permalink to this definition">¶</a></dt>
<dd><p>Get average counts per variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>libraries</strong> (<em>{'all'</em><em>, </em><em>'all_only'</em><em>, </em><em>list}</em>) – Include all libraries including a merge, only a merge of all
libraries, or a list of libraries.</p></li>
<li><p><strong>samples</strong> (<em>{'all'</em><em>, </em><em>list}</em>) – Include all samples or just samples in list.</p></li>
<li><p><strong>min_support</strong> (<em>int</em>) – Only include variants with variant call support &gt;= this.</p></li>
<li><p><strong>sample_rename</strong> (<em>dict</em><em> or </em><em>None</em>) – Rename samples by specifying original name as key and new name
as value.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Gives average counts per variant for each library and sample.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.classifyVariants">
<em class="property">static </em><code class="sig-name descname">classifyVariants</code><span class="sig-paren">(</span><em class="sig-param">df</em>, <em class="sig-param">*</em>, <em class="sig-param">variant_class_col='variant_class'</em>, <em class="sig-param">max_aa=2</em>, <em class="sig-param">syn_as_wt=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.classifyVariants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.classifyVariants" title="Permalink to this definition">¶</a></dt>
<dd><p>Classifies codon variants in <cite>df</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Must have columns named ‘aa_substitutions’, ‘n_aa_substitutions’,
and ‘n_codon_substitutions’. Such a data frame can be obtained via
<cite>variant_count_df</cite> or <cite>barcode_variant_df</cite> attributes of a
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> or via
<a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.func_scores" title="dms_variants.codonvarianttable.CodonVariantTable.func_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.func_scores()</span></code></a>.</p></li>
<li><p><strong>variant_class_col</strong> (<em>str</em>) – Name of column added to <cite>df</cite> that contains variant classification.
Overwritten if already exists.</p></li>
<li><p><strong>max_aa</strong> (<em>int</em>) – When classifying, group all with &gt;= this many amino-acid mutations.</p></li>
<li><p><strong>syn_as_wt</strong> (<em>bool</em>) – Do not have a category of ‘synonymous’ and instead classify
synonymous variants as ‘wildtype’. If using this option, <cite>df</cite>
does not need column ‘n_codon_substitutions’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>Copy of <cite>df</cite> with column specified by <cite>variant_class_col</cite> as:</dt><dd><ul class="simple">
<li><p>’wildtype’: no codon mutations</p></li>
<li><p>’synonymous’: only synonymous codon mutations</p></li>
<li><p>’stop’: at least one stop-codon mutation</p></li>
<li><p>’{n_aa} nonsynonymous’ where <cite>n_aa</cite> is number of amino-acid
mutations, or is ‘&gt;{max_aa}’ if more than <cite>max_aa</cite>.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
<span class="gp">... </span>        <span class="p">[(</span><span class="s1">&#39;AAA&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;AAG&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;ATA&#39;</span><span class="p">,</span> <span class="s1">&#39;M1* G5K&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;GAA&#39;</span><span class="p">,</span> <span class="s1">&#39;G5H&#39;</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;CTT&#39;</span><span class="p">,</span> <span class="s1">&#39;M1C G5C&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">(</span><span class="s1">&#39;CTT&#39;</span><span class="p">,</span> <span class="s1">&#39;M1A L3T G5C&#39;</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
<span class="gp">... </span>         <span class="p">],</span>
<span class="gp">... </span>        <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                 <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;n_codon_substitutions&#39;</span><span class="p">]</span>
<span class="gp">... </span>        <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_classify</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">all</span><span class="p">(</span><span class="n">df_classify</span><span class="o">.</span><span class="n">columns</span> <span class="o">==</span> <span class="p">[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                            <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                            <span class="s1">&#39;n_codon_substitutions&#39;</span><span class="p">,</span>
<span class="gp">... </span>                            <span class="s1">&#39;variant_class&#39;</span><span class="p">])</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_classify</span><span class="p">[[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_class&#39;</span><span class="p">]]</span>
<span class="go">  barcode     variant_class</span>
<span class="go">0     AAA          wildtype</span>
<span class="go">1     AAG        synonymous</span>
<span class="go">2     ATA              stop</span>
<span class="go">3     GAA   1 nonsynonymous</span>
<span class="go">4     CTT  &gt;1 nonsynonymous</span>
<span class="go">5     CTT  &gt;1 nonsynonymous</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_syn_as_wt</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span><span class="n">df</span><span class="p">,</span>
<span class="gp">... </span>                                                  <span class="n">syn_as_wt</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df_syn_as_wt</span><span class="p">[[</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;variant_class&#39;</span><span class="p">]]</span>
<span class="go">  barcode     variant_class</span>
<span class="go">0     AAA          wildtype</span>
<span class="go">1     AAG          wildtype</span>
<span class="go">2     ATA              stop</span>
<span class="go">3     GAA   1 nonsynonymous</span>
<span class="go">4     CTT  &gt;1 nonsynonymous</span>
<span class="go">5     CTT  &gt;1 nonsynonymous</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.codonToAAMuts">
<em class="property">classmethod </em><code class="sig-name descname">codonToAAMuts</code><span class="sig-paren">(</span><em class="sig-param">codon_mut_str</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.codonToAAMuts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.codonToAAMuts" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert string of codon mutations to amino-acid mutations.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>codon_mut_str</strong> (<em>str</em>) – Codon mutations, delimited by a space and in 1, … numbering.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Amino acid mutations in 1, … numbering.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">codonToAAMuts</span><span class="p">(</span><span class="s1">&#39;ATG1GTG GGA2GGC TGA3AGA&#39;</span><span class="p">)</span>
<span class="go">&#39;M1V *3R&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.from_variant_count_df">
<em class="property">classmethod </em><code class="sig-name descname">from_variant_count_df</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">variant_count_df_file</em>, <em class="sig-param">geneseq</em>, <em class="sig-param">drop_all_libs=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.from_variant_count_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.from_variant_count_df" title="Permalink to this definition">¶</a></dt>
<dd><p><a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a> from CSV of <cite>variant_count_df</cite>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Use this method when you have written a CSV file of <cite>variant_count_df</cite>
attribute of a <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>, and wish to re-initialize.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_count_df_file</strong> (<em>str</em>) – Name of CSV file containing the <cite>variant_count_df</cite>. Must have
following columns: “barcode”, “library”, “variant_call_support”,
“codon_substitutions”, “sample”, and “count”.</p></li>
<li><p><strong>geneseq</strong> (<em>str</em>) – Sequence of wildtype protein-coding gene.</p></li>
<li><p><strong>drop_all_libs</strong> (<em>bool</em>) – If there is a library named “all libraries”, drop it as it probably
added by <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.addMergedLibraries" title="dms_variants.codonvarianttable.CodonVariantTable.addMergedLibraries"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.addMergedLibraries()</span></code></a> and
duplicates information for the individual libraries.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.func_scores">
<code class="sig-name descname">func_scores</code><span class="sig-paren">(</span><em class="sig-param">preselection</em>, <em class="sig-param">*</em>, <em class="sig-param">pseudocount=0.5</em>, <em class="sig-param">by='barcode'</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">syn_as_wt=False</em>, <em class="sig-param">logbase=2</em>, <em class="sig-param">permit_zero_wt=False</em>, <em class="sig-param">permit_self_comp=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.func_scores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.func_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Get data frame with functional scores for variants.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The functional score is calculated from the change in counts
for a variant pre- and post-selection using the formula in
<a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et al (2018)</a>.</p>
<p>Specifically, let <span class="math notranslate nohighlight">\(n^v_{pre}\)</span> and <span class="math notranslate nohighlight">\(n^v_{post}\)</span> be
the counts of variant <span class="math notranslate nohighlight">\(v\)</span> pre- and post-selection, and
let <span class="math notranslate nohighlight">\(n^{wt}_{pre}\)</span> and <span class="math notranslate nohighlight">\(n^{wt}_{post}\)</span> be
the summed counts of <strong>all</strong> wildtype variants pre- and post-
selection.</p>
<p>Then the functional score of the variant is:</p>
<div class="math notranslate nohighlight">
\[f_v = \log_b\left(\frac{n^v_{post} /
n^{wt}_{post}}{n^v_{pre} / n^{wt}_{pre}}\right).\]</div>
<p>The variance due to Poisson sampling statistics is:</p>
<div class="math notranslate nohighlight">
\[\sigma^2_v = \frac{1}{\left(\ln b\right)^2}
\left(\frac{1}{n^v_{post}} + \frac{1}{n^{wt}_{post}} +
\frac{1}{n^v_{pre}} + \frac{1}{n^{wt}_{pre}}\right)\]</div>
<p>where <span class="math notranslate nohighlight">\(b\)</span> is logarithm base (see <cite>logbase</cite> parameter).</p>
<p>For both calculations, a pseudocount (see <cite>pseudocount</cite> parameter)
is added to each count first. The wildtype counts are computed
across all <strong>fully wildtype</strong> variants (see <cite>syn_as_wt</cite> for
how this is defined).</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>preselection</strong> (<em>str</em><em> or </em><em>dict</em>) – Pre-selection sample. If the same for all post-selection then
provide the name as str. If it differs among post-selection
samples, then provide a dict keyed by each post-selection
sample with the pre-selection sample being the value.</p></li>
<li><p><strong>pseudocount</strong> (<em>float</em>) – Pseudocount added to each count.</p></li>
<li><p><strong>by</strong> (<em>{'barcode'</em><em>, </em><em>'aa_substitutions'</em><em>, </em><em>'codon_substitutions'}</em>) – Compute effects for each barcode”, set of amino-acid substitutions,
or set of codon substitutions. In the last two cases, all barcodes
with each set of substitutions are combined (see <cite>combine_libs</cite>).
If you use “aa_substitutions” then it may be more sensible to set
<cite>syn_as_wt</cite> to <cite>True</cite>.</p></li>
<li><p><strong>syn_as_wt</strong> (<em>bool</em>) – In formula for functional scores, consider variants with only
synonymous mutations when determining wildtype counts? If <cite>False</cite>,
only variants with <strong>no</strong> mutations of any type contribute.</p></li>
<li><p><strong>libraries</strong> (<em>{'all'</em><em>, </em><em>'all_only'</em><em>, </em><em>list}</em>) – Perform calculation for all libraries including a merge
(named “all libraries”), only for the merge of all libraries,
or for the libraries in the list. If <cite>by</cite> is ‘barcode’, then
the barcodes for the merge have the library name pre-pended.</p></li>
<li><p><strong>logbase</strong> (<em>float</em>) – Base for logarithm when calculating functional score.</p></li>
<li><p><strong>permit_zero_wt</strong> (<em>bool</em>) – If the wildtype counts are zero for any sample, raise an error
or permit the calculation to proceed just using pseudocount?</p></li>
<li><p><strong>permit_self_comp</strong> (<em>bool</em>) – Permit comparisons for sample pre- and post-selection?</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>Has the following columns:</dt><dd><ul class="simple">
<li><p>”library”: the library</p></li>
<li><p>”pre_sample”: the pre-selection sample</p></li>
<li><p>”post_sample”: the post-selection sample</p></li>
<li><p>value corresponding to grouping used to compute effects (<cite>by</cite>)</p></li>
<li><p>”func_score”: the functional score</p></li>
<li><p>”func_score_var”: variance on the functional score</p></li>
<li><p>”pre_count”: pre-selection counts</p></li>
<li><p>”post_count: post-selection counts</p></li>
<li><p>”pre_count_wt”: pre-selection counts for all wildtype</p></li>
<li><p>”post_count_wt”: post-selection counts for all wildtype</p></li>
<li><p>”pseudocount”: the pseudocount value</p></li>
<li><p>as many of “aa_substitutions”, “n_aa_substitutions”,
“codon_substitutions”, and “n_codon_substitutions”
as can be retained given value of <cite>by</cite>.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.mutCounts">
<code class="sig-name descname">mutCounts</code><span class="sig-paren">(</span><em class="sig-param">variant_type</em>, <em class="sig-param">mut_type</em>, <em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.mutCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.mutCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Get counts of each individual mutation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include just single mutants, or all mutants?</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Tidy data frame with columns named “library”, “sample”,
“mutation”, “count”, “mutation_type”, and “site”.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.n_variants_df">
<code class="sig-name descname">n_variants_df</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">variant_type='all'</em>, <em class="sig-param">mut_type=None</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.n_variants_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.n_variants_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Get number of variants per library / sample.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include all variants or just those with &lt;= 1 <cite>mut_type</cite> mutation.</p></li>
<li><p><strong>mut_type</strong> (<em>{'aa'</em><em>, </em><em>'codon'</em><em>, </em><em>None}</em>) – If <cite>variant_type</cite> is ‘single’, indicate what type of single
mutants we are filtering for.</p></li>
<li><p><strong>other args</strong> (<em>All</em>) – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.numCodonMutsByType">
<code class="sig-name descname">numCodonMutsByType</code><span class="sig-paren">(</span><em class="sig-param">variant_type</em>, <em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.numCodonMutsByType"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.numCodonMutsByType" title="Permalink to this definition">¶</a></dt>
<dd><p>Get average nonsynonymous, synonymous, stop mutations per variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>all_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumCodonMutsByType()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Data frame with average mutations of each type per variant.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.plotAvgCountsPerVariant">
<code class="sig-name descname">plotAvgCountsPerVariant</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">plotfile=None</em>, <em class="sig-param">orientation='h'</em>, <em class="sig-param">widthscale=1</em>, <em class="sig-param">heightscale=1</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotAvgCountsPerVariant"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotAvgCountsPerVariant" title="Permalink to this definition">¶</a></dt>
<dd><p>Get average counts per variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>libraries</strong> (<em>{'all'</em><em>, </em><em>'all_only'</em><em>, </em><em>list}</em>) – Include all libraries including a merge, only a merge of all
libraries, or a list of libraries.</p></li>
<li><p><strong>samples</strong> (<em>{'all'</em><em>, </em><em>list}</em>) – Include all samples or just samples in list.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage">
<code class="sig-name descname">plotCumulMutCoverage</code><span class="sig-paren">(</span><em class="sig-param">variant_type</em>, <em class="sig-param">mut_type</em>, <em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">plotfile=None</em>, <em class="sig-param">orientation='h'</em>, <em class="sig-param">widthscale=1</em>, <em class="sig-param">heightscale=1</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">max_count=None</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotCumulMutCoverage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="Permalink to this definition">¶</a></dt>
<dd><p>Fraction of mutations seen &lt;= some number of times.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include all variants or just those with &lt;=1 <cite>mut_type</cite> mutation.</p></li>
<li><p><strong>max_count</strong> (<em>None</em><em> or </em><em>int</em>) – Plot cumulative fraction plot out to this number of observations.
If <cite>None</cite>, a reasonable value is automatically determined.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.plotCumulVariantCounts">
<code class="sig-name descname">plotCumulVariantCounts</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">variant_type='all'</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">plotfile=None</em>, <em class="sig-param">orientation='h'</em>, <em class="sig-param">widthscale=1</em>, <em class="sig-param">heightscale=1</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">mut_type='aa'</em>, <em class="sig-param">tot_variants_hline=True</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotCumulVariantCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulVariantCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot number variants with &gt;= that each number of counts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include all variants or just those with &lt;=1 <cite>mut_type</cite> mutation.</p></li>
<li><p><strong>tot_variants_hline</strong> (<em>bool</em>) – Include dotted horizontal line indicating total number of variants.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.plotMutFreqs">
<code class="sig-name descname">plotMutFreqs</code><span class="sig-paren">(</span><em class="sig-param">variant_type</em>, <em class="sig-param">mut_type</em>, <em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">plotfile=None</em>, <em class="sig-param">orientation='h'</em>, <em class="sig-param">widthscale=1</em>, <em class="sig-param">heightscale=1</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotMutFreqs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotMutFreqs" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutation frequency along length of gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>parameters</strong> (<em>All</em>) – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotCumulMutCoverage()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.plotMutHeatmap">
<code class="sig-name descname">plotMutHeatmap</code><span class="sig-paren">(</span><em class="sig-param">variant_type</em>, <em class="sig-param">mut_type</em>, <em class="sig-param">*</em>, <em class="sig-param">count_or_frequency='frequency'</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">plotfile=None</em>, <em class="sig-param">orientation='h'</em>, <em class="sig-param">widthscale=1</em>, <em class="sig-param">heightscale=1</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotMutHeatmap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotMutHeatmap" title="Permalink to this definition">¶</a></dt>
<dd><p>Heatmap of mutation counts or frequencies.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>count_or_frequency</strong> (<em>{'count'</em><em>, </em><em>'frequency'}</em>) – Plot mutation counts or frequencies?</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage" title="dms_variants.codonvarianttable.CodonVariantTable.plotCumulMutCoverage"><code class="xref py py-meth docutils literal notranslate"><span class="pre">CodonVariantTable.plotCumulMutCoverage()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType">
<code class="sig-name descname">plotNumCodonMutsByType</code><span class="sig-paren">(</span><em class="sig-param">variant_type</em>, <em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">plotfile=None</em>, <em class="sig-param">orientation='h'</em>, <em class="sig-param">widthscale=1</em>, <em class="sig-param">heightscale=1</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">ylabel=None</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotNumCodonMutsByType"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumCodonMutsByType" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot average nonsynonymous, synonymous, stop mutations per variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variant_type</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – Include all variants or just those with &lt;=1 codon mutation.</p></li>
<li><p><strong>ylabel</strong> (<em>None</em><em> or </em><em>str</em>) – If not <cite>None</cite>, specify y-axis label (otherwise it is autoset).</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram">
<code class="sig-name descname">plotNumMutsHistogram</code><span class="sig-paren">(</span><em class="sig-param">mut_type</em>, <em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">samples='all'</em>, <em class="sig-param">plotfile=None</em>, <em class="sig-param">orientation='h'</em>, <em class="sig-param">widthscale=1</em>, <em class="sig-param">heightscale=1</em>, <em class="sig-param">min_support=1</em>, <em class="sig-param">max_muts=None</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotNumMutsHistogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot histograms of number of mutations per variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mut_type</strong> (<em>{'codon'</em><em> or </em><em>'aa'}</em>) – Mutation type.</p></li>
<li><p><strong>libraries</strong> (<em>{'all'</em><em>, </em><em>'all_only'</em><em>, </em><em>list}</em>) – Include all libraries including a merge, only a merge of all
libraries, or a list of libraries.</p></li>
<li><p><strong>samples</strong> (<em>{'all'</em><em>, </em><em>None</em><em>, </em><em>list}</em>) – Include all samples, a list of samples, or <cite>None</cite> to just count
each barcoded variant once.</p></li>
<li><p><strong>plotfile</strong> (<em>None</em><em> or </em><em>str</em>) – Name of file to which to save plot.</p></li>
<li><p><strong>orientation</strong> (<em>{'h'</em><em>, </em><em>'v'}</em>) – Facet libraries horizontally or vertically?</p></li>
<li><p><strong>widthscale</strong> (<em>float</em>) – Expand width of plot by this factor.</p></li>
<li><p><strong>heightscale</strong> (<em>float</em>) – Expand height of plot by this factor.</p></li>
<li><p><strong>min_support</strong> (<em>int</em>) – Only include variants with variant call support &gt;= this.</p></li>
<li><p><strong>max_muts</strong> (<em>int</em><em> or </em><em>None</em>) – Group together all variants with &gt;= this many mutations.</p></li>
<li><p><strong>sample_rename</strong> (<em>dict</em><em> or </em><em>None</em>) – Rename samples by specifying original name as key and new name
as value.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.plotVariantSupportHistogram">
<code class="sig-name descname">plotVariantSupportHistogram</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">libraries='all'</em>, <em class="sig-param">plotfile=None</em>, <em class="sig-param">orientation='h'</em>, <em class="sig-param">widthscale=1</em>, <em class="sig-param">heightscale=1</em>, <em class="sig-param">max_support=None</em>, <em class="sig-param">sample_rename=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.plotVariantSupportHistogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.plotVariantSupportHistogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot histogram of variant call support for variants.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>max_support</strong> (<em>int</em><em> or </em><em>None</em>) – Group together all variants with &gt;= this support.</p></li>
<li><p><strong>other_parameters</strong> – Same as for <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram" title="dms_variants.codonvarianttable.CodonVariantTable.plotNumMutsHistogram"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable.plotNumMutsHistogram</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.samples">
<code class="sig-name descname">samples</code><span class="sig-paren">(</span><em class="sig-param">library</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.samples"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.samples" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all samples for <cite>library</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>library</strong> (<em>str</em>) – Valid <cite>library</cite> for the <a class="reference internal" href="#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">CodonVariantTable</span></code></a>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>All samples for which barcode counts have been added.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.subs_to_seq">
<code class="sig-name descname">subs_to_seq</code><span class="sig-paren">(</span><em class="sig-param">subs</em>, <em class="sig-param">subs_type='codon'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.subs_to_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.subs_to_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert substitutions to full sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>subs</strong> (<em>str</em>) – Space delimited substitutions in 1, … numbering.</p></li>
<li><p><strong>subs_type</strong> (<em>{'codon'</em><em>, </em><em>'aa'}</em>) – Are substitutions codon or amino acid?</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Sequence created by these mutations, either codon or
amino acid depending on value of <cite>subs_type</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">geneseq</span> <span class="o">=</span> <span class="s1">&#39;ATGGGATGA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;library,barcode,substitutions,&#39;</span>
<span class="gp">... </span>                <span class="s1">&#39;variant_call_support&#39;</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">variants</span> <span class="o">=</span> <span class="n">CodonVariantTable</span><span class="p">(</span>
<span class="gp">... </span>                <span class="n">barcode_variant_file</span><span class="o">=</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
<span class="gp">... </span>                <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span>
<span class="gp">... </span>                <span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">subs_to_seq</span><span class="p">(</span><span class="s1">&#39;GGA2CGT ATG1GTG&#39;</span><span class="p">)</span>
<span class="go">&#39;GTGCGTTGA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">variants</span><span class="o">.</span><span class="n">subs_to_seq</span><span class="p">(</span><span class="s1">&#39;*3W M1C&#39;</span><span class="p">,</span> <span class="n">subs_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">)</span>
<span class="go">&#39;CGW&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.valid_barcodes">
<code class="sig-name descname">valid_barcodes</code><span class="sig-paren">(</span><em class="sig-param">library</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.valid_barcodes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.valid_barcodes" title="Permalink to this definition">¶</a></dt>
<dd><p>Set of valid barcodes for <cite>library</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>library</strong> (<em>str</em>) – Name of a valid library.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>set</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.codonvarianttable.CodonVariantTable.writeCodonCounts">
<code class="sig-name descname">writeCodonCounts</code><span class="sig-paren">(</span><em class="sig-param">single_or_all</em>, <em class="sig-param">*</em>, <em class="sig-param">outdir=None</em>, <em class="sig-param">include_all_libs=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/codonvarianttable.html#CodonVariantTable.writeCodonCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.codonvarianttable.CodonVariantTable.writeCodonCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Write codon counts files for all libraries and samples.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Useful if you want to analyze individual mutations using
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/">dms_tools2</a>. File format:
<a class="reference external" href="https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html#counts-file">https://jbloomlab.github.io/dms_tools2/dms2_bcsubamp.html#counts-file</a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>single_or_all</strong> (<em>{'single'</em><em>, </em><em>'all'}</em>) – If ‘single’, then counts just from single-codon mutants and
wildtype, and count all wildtype codons and just mutated codons
for single-codon mutants. If ‘all’, count all codons for all
variants at all sites. Appropriate if enrichment of each mutation
is supposed to represent its effect for ‘single’, and if enrichment
of mutation is supposed to represent its average effect across
genetic backgrounds in the library for ‘all’ provided mutations are
Poisson distributed.</p></li>
<li><p><strong>outdir</strong> (<em>None</em><em> or </em><em>str</em>) – Name of directory to write counts, created if it does not exist.
Use <cite>None</cite> to write to current directory.</p></li>
<li><p><strong>include_all_libs</strong> (<em>bool</em>) – Include data for a library (named “all-libraries”) that has
summmed data for all individual libraries if multiple libraries.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Gives names of created files. Has columns “library”, “sample”,
and “countfile”. The “countfile” columns gives name of the
created CSV file, <code class="docutils literal notranslate"><span class="pre">&lt;library&gt;_&lt;sample&gt;_codoncounts.csv</span></code>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_variants">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">dms_variants documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dms_variants.html">dms_variants package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dms_variants.html#dms-variants">dms_variants</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="dms_variants.html#submodules">Submodules</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="dms_variants.html">dms_variants package</a><ul>
      <li>Previous: <a href="dms_variants.binarymap.html" title="previous chapter">binarymap</a></li>
      <li>Next: <a href="dms_variants.constants.html" title="next chapter">constants</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2019.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms_variants.codonvarianttable.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>