
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>simulate &#8212; dms_variants 0.3.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="utils" href="dms_variants.utils.html" />
    <link rel="prev" title="plotnine_themes" href="dms_variants.plotnine_themes.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-dms_variants.simulate"></span><div class="section" id="simulate">
<h1>simulate<a class="headerlink" href="#simulate" title="Permalink to this headline">¶</a></h1>
<p>Simulate data.</p>
<dl class="class">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator">
<em class="property">class </em><code class="sig-prename descclassname">dms_variants.simulate.</code><code class="sig-name descname">SigmoidPhenotypeSimulator</code><span class="sig-paren">(</span><em class="sig-param">geneseq</em>, <em class="sig-param">*</em>, <em class="sig-param">seed=1</em>, <em class="sig-param">wt_latent=5</em>, <em class="sig-param">norm_weights=((0.4</em>, <em class="sig-param">-0.7</em>, <em class="sig-param">1.5)</em>, <em class="sig-param">(0.6</em>, <em class="sig-param">-7</em>, <em class="sig-param">3.5))</em>, <em class="sig-param">stop_effect=-15</em>, <em class="sig-param">min_observed_enrichment=0.001</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#SigmoidPhenotypeSimulator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Simulate phenotypes under sigmoid global epistasis model.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Mutational effects on latent phenotype are simulated to follow
compound normal distribution; latent phenotype maps to observed
enrichment via sigmoid, and the observed phenotype is the
<span class="math notranslate nohighlight">\(\log_2\)</span> of the enrichment. This distinction between latent and
observed phenotype parallels the “global epistasis” models of
<a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et al</a> and
<a class="reference external" href="http://www.genetics.org/content/205/3/1079">Sailer and Harms</a>.</p>
<p>Specifically, let <span class="math notranslate nohighlight">\(p_{\rm{latent}}\)</span> be the latent phenotype
of a variant, let <span class="math notranslate nohighlight">\(p_{\rm{observed}}\)</span> be the observed phenotype,
and let <span class="math notranslate nohighlight">\(E_{\rm{observed}}\)</span> be the observed enrichment. Also
let <span class="math notranslate nohighlight">\(p_{\rm{latent, wt}}\)</span> be the latent phenotype of the
wildtype. Then:</p>
<div class="math notranslate nohighlight">
\[\begin{split}E_{\rm{observed}}
&amp;=&amp; \frac{\left(1 + \exp\left(-p_{\rm{latent, wt}}\right)\right)
          \left(1 - E_{\rm{min}}\right)}
         {1 + \exp\left(-p_{\rm{latent}}\right)} +
         E_{\rm{min}} \\
p_{\rm{observed}} &amp;=&amp; \log_2 E_{\rm{observed}}\end{split}\]</div>
<p>where <span class="math notranslate nohighlight">\(0 \le E_{\rm{min}} &lt; 1\)</span> is the minimum possible observed
enrichment. Typically <span class="math notranslate nohighlight">\(E_{\rm{min}} &gt; 0\)</span> in real experiments
as pseudocounts are used so estimates of enrichments can’t be zero.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>geneseq</strong> (<em>str</em>) – Codon sequence of wild-type gene.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em> or </em><em>None</em>) – Random number seed.</p></li>
<li><p><strong>wt_latent</strong> (<em>float</em>) – Latent phenotype of wildtype.</p></li>
<li><p><strong>norm_weights</strong> (<em>list</em><em> or </em><em>tuple of tuples</em>) – Specify compound normal distribution of mutational effects on
latent phenotype as <cite>(weight, mean, sd)</cite> for each Gaussian.</p></li>
<li><p><strong>stop_effect</strong> (<em>float</em>) – Effect of stop codon at any position.</p></li>
<li><p><strong>min_observed_enrichment</strong> (<em>float</em>) – Minimum possible observed enrichment.</p></li>
</ul>
</dd>
</dl>
<dl class="attribute">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator.wt_latent">
<code class="sig-name descname">wt_latent</code><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.wt_latent" title="Permalink to this definition">¶</a></dt>
<dd><p>Wildtype latent phenotype.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator.muteffects">
<code class="sig-name descname">muteffects</code><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.muteffects" title="Permalink to this definition">¶</a></dt>
<dd><p>Effect on latent phenotype of each amino-acid mutation.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator.latentPhenotype">
<code class="sig-name descname">latentPhenotype</code><span class="sig-paren">(</span><em class="sig-param">subs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#SigmoidPhenotypeSimulator.latentPhenotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.latentPhenotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Latent phenotype of a variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>subs</strong> (<em>str</em>) – Space-delimited list of amino-acid substitutions.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Latent phenotype of variant.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator.latentToObserved">
<code class="sig-name descname">latentToObserved</code><span class="sig-paren">(</span><em class="sig-param">latent</em>, <em class="sig-param">value</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#SigmoidPhenotypeSimulator.latentToObserved"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.latentToObserved" title="Permalink to this definition">¶</a></dt>
<dd><p>Observed enrichment or observed phenotype from latent phenotype.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>latent</strong> (<em>float</em><em> or </em><em>numpy.ndarray</em>) – The latent phenotype.</p></li>
<li><p><strong>value</strong> (<em>{'enrichment'</em><em>, </em><em>'phenotype'}</em>) – Get the observed enrichment or the observed phenotype?</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The observed enrichment or phenotype.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float or numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator.observedEnrichment">
<code class="sig-name descname">observedEnrichment</code><span class="sig-paren">(</span><em class="sig-param">subs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#SigmoidPhenotypeSimulator.observedEnrichment"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.observedEnrichment" title="Permalink to this definition">¶</a></dt>
<dd><p>Observed enrichment of a variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>subs</strong> (<em>str</em>) – Space-delimited list of amino-acid substitutions.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Observed enrichment relative to wildtype.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator.observedPhenotype">
<code class="sig-name descname">observedPhenotype</code><span class="sig-paren">(</span><em class="sig-param">subs</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#SigmoidPhenotypeSimulator.observedPhenotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.observedPhenotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Observed phenotype of a variant.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>subs</strong> (<em>str</em>) – Space-delimited list of amino-acid substitutions.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Observed phenotype of variant.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator.plotLatentVsObserved">
<code class="sig-name descname">plotLatentVsObserved</code><span class="sig-paren">(</span><em class="sig-param">value</em>, <em class="sig-param">*</em>, <em class="sig-param">latent_min=-15</em>, <em class="sig-param">latent_max=10</em>, <em class="sig-param">npoints=200</em>, <em class="sig-param">wt_vline=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#SigmoidPhenotypeSimulator.plotLatentVsObserved"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.plotLatentVsObserved" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot observed enrichment/phenotype as function of latent phenotype.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>{'enrichment'</em><em>, </em><em>'phenotype'}</em>) – Do we plot observed enrichment or observed phenotype?</p></li>
<li><p><strong>latent_min</strong> (<em>float</em>) – Smallest value of latent phenotype on plot.</p></li>
<li><p><strong>latent_max</strong> (<em>float</em>) – Largest value of latent phenotype on plot.</p></li>
<li><p><strong>npoints</strong> (<em>int</em>) – Plot a line fit to this many points.</p></li>
<li><p><strong>wt_vline</strong> (<em>bool</em>) – Draw a vertical line at the wildtype latent phenotype.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Plot of observed enrichment or phenotype as function of latent
phenotype.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="dms_variants.simulate.SigmoidPhenotypeSimulator.plotMutsHistogram">
<code class="sig-name descname">plotMutsHistogram</code><span class="sig-paren">(</span><em class="sig-param">value</em>, <em class="sig-param">*</em>, <em class="sig-param">mutant_order=1</em>, <em class="sig-param">bins=30</em>, <em class="sig-param">wt_vline=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#SigmoidPhenotypeSimulator.plotMutsHistogram"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.plotMutsHistogram" title="Permalink to this definition">¶</a></dt>
<dd><p>Plot distribution of phenotype for all mutants of a given order.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>value</strong> (<em>{'latentPhenotype'</em><em>, </em><em>'observedPhenotype'</em><em>, </em><em>'observedEnrichment'}</em>) – What value to plot.</p></li>
<li><p><strong>mutant_order</strong> (<em>int</em>) – Plot mutations of this order. Currently only works for 1
(single mutants).</p></li>
<li><p><strong>bins</strong> (<em>int</em>) – Number of bins in histogram.</p></li>
<li><p><strong>wt_vline</strong> (<em>bool</em>) – Draw a vertical line at the wildtype latent phenotype.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Histogram of phenotype for all mutants.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>plotnine.ggplot.ggplot</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="dms_variants.simulate.mutate_seq">
<code class="sig-prename descclassname">dms_variants.simulate.</code><code class="sig-name descname">mutate_seq</code><span class="sig-paren">(</span><em class="sig-param">seq</em>, <em class="sig-param">nmuts</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#mutate_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.mutate_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Mutate a nucleotide sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seq</strong> (<em>str</em>) – Sequence to mutate.</p></li>
<li><p><strong>nmuts</strong> (<em>int</em>) – Number of mutations to make to sequence.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Mutagenized sequence, all upper-case.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Seed <cite>random.seed</cite> for reproducible output.</p>
</div>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mutate_seq</span><span class="p">(</span><span class="s1">&#39;ATGCAA&#39;</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="go">&#39;AAGCGA&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_variants.simulate.rand_seq">
<code class="sig-prename descclassname">dms_variants.simulate.</code><code class="sig-name descname">rand_seq</code><span class="sig-paren">(</span><em class="sig-param">seqlen</em>, <em class="sig-param">*</em>, <em class="sig-param">exclude=None</em>, <em class="sig-param">nseqs=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#rand_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.rand_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Random nucleotide sequence(s).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>seqlen</strong> (<em>int</em>) – Length of sequence(s).</p></li>
<li><p><strong>exclude</strong> (<em>None</em><em> or </em><em>list</em>) – Do not generate any of the sequences in this list.</p></li>
<li><p><strong>nseqs</strong> (<em>int</em>) – If &gt;1, return a list of this many sequences.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A random sequence or list of them depending on value of <cite>nseqs</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>str or list</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Seed <cite>random.seed</cite> for reproducible output.</p>
</div>
<p class="rubric">Example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rand_seq</span><span class="p">(</span><span class="mi">4</span><span class="p">)</span>
<span class="go">&#39;CAGA&#39;</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rand_seq</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">nseqs</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">[&#39;CAGAT&#39;, &#39;TTTCA&#39;, &#39;TATTA&#39;]</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rand_seq</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="n">nseqs</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">exclude</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;TTTCA&#39;</span><span class="p">])</span>
<span class="go">[&#39;CAGAT&#39;, &#39;TATTA&#39;, &#39;TGCAG&#39;]</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="dms_variants.simulate.simulateSampleCounts">
<code class="sig-prename descclassname">dms_variants.simulate.</code><code class="sig-name descname">simulateSampleCounts</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">variants</em>, <em class="sig-param">phenotype_func</em>, <em class="sig-param">variant_error_rate</em>, <em class="sig-param">pre_sample</em>, <em class="sig-param">post_samples</em>, <em class="sig-param">pre_sample_name='pre-selection'</em>, <em class="sig-param">seed=1</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#simulateSampleCounts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.simulateSampleCounts" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate pre- and post-selection variant counts.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Add to <a class="reference internal" href="dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.codonvarianttable.CodonVariantTable</span></code></a> using
<a class="reference internal" href="dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts" title="dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts"><code class="xref py py-meth docutils literal notranslate"><span class="pre">dms_variants.codonvarianttable.CodonVariantTable.addSampleCounts()</span></code></a></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>variants</strong> (class:<cite>dms_variants.codonvarianttable.CodonVariantTable</cite>) – Holds variants used in simulation.</p></li>
<li><p><strong>phenotype_func</strong> (<em>function</em>) – Takes a space-delimited string of amino-acid substitutions and returns
a number giving enrichment of variant relative to wildtype. For
instance, <a class="reference internal" href="#dms_variants.simulate.SigmoidPhenotypeSimulator.observedEnrichment" title="dms_variants.simulate.SigmoidPhenotypeSimulator.observedEnrichment"><code class="xref py py-meth docutils literal notranslate"><span class="pre">SigmoidPhenotypeSimulator.observedEnrichment()</span></code></a>.</p></li>
<li><p><strong>variant_error_rate</strong> (<em>float</em>) – Rate at which variants mis-called. Provide the probability that a
variant has a spuriously called (or missing) codon mutation; each
variant then has a random codon mutation added or removed with this
probability before being passed to <cite>phenotype_func</cite>.</p></li>
<li><p><strong>pre_sample</strong> (<em>pandas.DataFrame</em><em> or </em><em>dict</em>) – Counts of each variant pre-selection. To specify counts, provide
data frame with columns “library”, “barcode”, and “count”. To
simulate, provide dict with keys “total_count” and “uniformity”,
and for each library, we simulate pre-selection counts as a draw
of “total_count” counts from a multinomial parameterized by pre-
selection frequencies drawn from a Dirichlet distribution with
concentration parameter of “uniformity” (5 is reasonable value).</p></li>
<li><p><strong>post_samples</strong> (<em>dict</em>) – Keyed by name of each sample with value another dict keyed by
‘total_count’, ‘noise’, and ‘bottleneck’. Counts drawn from
multinomial parameterized by pre-selection frerquencies after
passing through bottleneck of indicated size, and adding noise
by mutiplying phenotype by a random variable with mean 1 and
standard deviation specified by ‘noise’ (0 is no noise).</p></li>
<li><p><strong>pre_sample_name</strong> (<em>str</em>) – Name used for the pre-selection sample.</p></li>
<li><p><strong>seed</strong> (<em>None</em><em> or </em><em>int</em>) – If not <cite>None</cite>, random number seed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>Data frame with the following columns:</dt><dd><ul class="simple">
<li><p>”library”</p></li>
<li><p>”barcode”</p></li>
<li><p>”sample”</p></li>
<li><p>”count”</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="dms_variants.simulate.simulate_CodonVariantTable">
<code class="sig-prename descclassname">dms_variants.simulate.</code><code class="sig-name descname">simulate_CodonVariantTable</code><span class="sig-paren">(</span><em class="sig-param">*</em>, <em class="sig-param">geneseq</em>, <em class="sig-param">bclen</em>, <em class="sig-param">library_specs</em>, <em class="sig-param">seed=1</em>, <em class="sig-param">variant_call_support=0.5</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/dms_variants/simulate.html#simulate_CodonVariantTable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#dms_variants.simulate.simulate_CodonVariantTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate <a class="reference internal" href="dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.codonvarianttable.CodonVariantTable</span></code></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only simulates the variants, not counts for samples. To simulate counts,
use <a class="reference internal" href="#dms_variants.simulate.simulateSampleCounts" title="dms_variants.simulate.simulateSampleCounts"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulateSampleCounts()</span></code></a>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>geneseq</strong> (<em>str</em>) – Sequence of wildtype protein-coding gene.</p></li>
<li><p><strong>bclen</strong> (<em>int</em>) – Length of the barcodes; must enable complexity at least
10-fold greater than max number of variants.</p></li>
<li><p><strong>library_specs</strong> (<em>dict</em>) – Specifications for each simulated library. Keys are ‘avgmuts’
and ‘nvariants’. Mutations per variant are Poisson distributed.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em> or </em><em>None</em>) – Random number seed or <cite>None</cite> to set no seed.</p></li>
<li><p><strong>variant_call_support</strong> (<em>int &gt; 0</em>) – Support is floor of 1 plus draw from exponential distribution
with this mean.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable" title="dms_variants.codonvarianttable.CodonVariantTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">dms_variants.codonvarianttable.CodonVariantTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_variants">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">dms_variants documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dms_variants.html">dms_variants package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="dms_variants.html#dms-variants">dms_variants</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="dms_variants.html#submodules">Submodules</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="dms_variants.html">dms_variants package</a><ul>
      <li>Previous: <a href="dms_variants.plotnine_themes.html" title="previous chapter">plotnine_themes</a></li>
      <li>Next: <a href="dms_variants.utils.html" title="next chapter">utils</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2019.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/dms_variants.simulate.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>