
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Fitting gobal epistasis models with experimental “noise” &#8212; dms_variants 0.4.6 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <script async="async" type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Predict unseen variants using global epistasis models" href="predict_variants.html" />
    <link rel="prev" title="Analyze simulated library of codon variants" href="codonvariant_sim_data.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput,
div.nbinput div.prompt,
div.nbinput div.input_area,
div.nbinput div[class*=highlight],
div.nbinput div[class*=highlight] pre,
div.nboutput,
div.nbinput div.prompt,
div.nbinput div.output_area,
div.nboutput div[class*=highlight],
div.nboutput div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput,
div.nboutput {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput,
    div.nboutput {
        flex-direction: column;
    }
}

/* input container */
div.nbinput {
    padding-top: 5px;
}

/* last container */
div.nblast {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput div.prompt,
div.nboutput div.prompt {
    min-width: 5ex;
    padding-top: 0.4em;
    padding-right: 0.4em;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput div.prompt,
    div.nboutput div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput div.prompt pre,
div.nboutput div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput div.input_area,
div.nboutput div.output_area {
    padding: 0.4em;
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput div.input_area,
    div.nboutput div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput div.math p {
    text-align: left;
}

/* standard error */
div.nboutput div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="Fitting-gobal-epistasis-models-with-experimental-“noise”">
<h1>Fitting gobal epistasis models with experimental “noise”<a class="headerlink" href="#Fitting-gobal-epistasis-models-with-experimental-“noise”" title="Permalink to this headline">¶</a></h1>
<p>The global epistasis models described by <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a> fit the data using a Gaussian-likelihood function that assumes that deviations from the models are normally distributed (see the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html">globalepistasis module documentation</a> for more details). Those models also accommodate estimates of the noise for each measurement, which are easy to make if the noise is due to
statistical processes that are simple to model (such as finite sequencing depth leading to sample statistics).</p>
<p>But often real deep mutational scanning experiments have other sources of experimental “noise” that are not easily modeled or even estimated. For instance, the sequencing used to determine the mutations present in a variance might occassionally wrong, leading to mis-calling of variant sequences. Or (more commonly in <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/27271655">Bloom lab influenza virus experiments</a>) there may be experimental bottlenecks in the number of variants that are passaged from the
pre-selection to post-selection conditions.</p>
<p>Here we analyze the effects of these sources of unmodeled statistic noise on the fitting of epistasis models to simulated data. The main conclusions are:</p>
<ol class="arabic simple">
<li><p>The Gaussian-likelihood global epistasis models work very well when there isn’t too much unmodeled experimental noise.</p></li>
<li><p>However, these models do become much less accurate when there is a lot of noise, probably due to the fact that very noisy measurements substantially skew the model fit.</p></li>
<li><p>An alternative way to calculate the likelihoods by assuming a Cauchy distribution of errors (see the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html">globalepistasis module documentation</a> for more details) works better when there is a lot of experimental noise, probably because the Cauchy distribution has fatter tails and so is less affected by outlier points.</p></li>
</ol>
<div class="section" id="Setup-for-analysis">
<h2>Setup for analysis<a class="headerlink" href="#Setup-for-analysis" title="Permalink to this headline">¶</a></h2>
<p>Import Python modules / packages:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">collections</span>
<span class="kn">import</span> <span class="nn">itertools</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">tempfile</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="kn">from</span> <span class="nn">plotnine</span> <span class="k">import</span> <span class="o">*</span>

<span class="kn">import</span> <span class="nn">scipy</span>

<span class="kn">import</span> <span class="nn">dms_variants.binarymap</span>
<span class="kn">import</span> <span class="nn">dms_variants.codonvarianttable</span>
<span class="kn">import</span> <span class="nn">dms_variants.globalepistasis</span>
<span class="kn">import</span> <span class="nn">dms_variants.plotnine_themes</span>
<span class="kn">import</span> <span class="nn">dms_variants.simulate</span>
<span class="kn">from</span> <span class="nn">dms_variants.constants</span> <span class="k">import</span> <span class="n">CBPALETTE</span><span class="p">,</span> <span class="n">CODONS_NOSTOP</span>
</pre></div>
</div>
</div>
<p>Set parameters that define the simulated data. We want to simulate some of the samples with a lot of “noise.” Here that noise takes the form of a very tight bottleneck when transferring the pre-selection library to the selection (such bottlenecks are unavoidable in some experiments). Specifically, we simulate experiments with: - A <em>loose bottleneck</em>: a lot of the pre-selection library is effectively transferred to the post-selection condition, meaning there is low “noise.” - A <em>mid bottleneck</em>:
there is an appreciable but not extreme bottleneck in transferring the pre-selection library to the post-selection condition, meaning there is appreciable but no extreme “noise.” - A <em>tight bottleneck</em>: there is a very narrow bottleneck when transferring the pre-selection library to the post-selection condition, leading to lots of “noise.”</p>
<p>The exact simulation parameters are defined below:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">seed</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># random number seed</span>
<span class="n">genelength</span> <span class="o">=</span> <span class="mi">40</span>  <span class="c1"># gene length in codons</span>
<span class="n">libs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;lib_1&#39;</span><span class="p">]</span>  <span class="c1"># distinct libraries of gene</span>
<span class="n">variants_per_lib</span> <span class="o">=</span> <span class="mi">500</span> <span class="o">*</span> <span class="n">genelength</span>  <span class="c1"># variants per library</span>
<span class="n">avgmuts</span> <span class="o">=</span> <span class="mf">2.0</span>  <span class="c1"># average codon mutations per variant</span>
<span class="n">bclen</span> <span class="o">=</span> <span class="mi">16</span>  <span class="c1"># length of nucleotide barcode for each variant</span>
<span class="n">variant_error_rate</span> <span class="o">=</span> <span class="mf">0.005</span>  <span class="c1"># rate at which variant sequence mis-called</span>
<span class="n">avgdepth_per_variant</span> <span class="o">=</span> <span class="mi">100</span>  <span class="c1"># average per-variant sequencing depth</span>
<span class="n">lib_uniformity</span> <span class="o">=</span> <span class="mi">5</span>  <span class="c1"># uniformity of library pre-selection</span>
<span class="n">noise</span> <span class="o">=</span> <span class="mf">0.02</span>  <span class="c1"># random noise in selections</span>
<span class="n">bottlenecks</span> <span class="o">=</span> <span class="p">{</span>  <span class="c1"># bottlenecks from pre- to post-selection</span>
        <span class="s1">&#39;tight_bottle&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span><span class="p">,</span>
        <span class="s1">&#39;mid_bottle&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span>
        <span class="s1">&#39;loose_bottle&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span>
        <span class="p">}</span>
</pre></div>
</div>
</div>
<p>Seed random number generator for reproducible output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="n">seed</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Set pandas display options to show large chunks of Data Frames in this example:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_columns&#39;</span><span class="p">,</span> <span class="mi">20</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.width&#39;</span><span class="p">,</span> <span class="mi">500</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Hide warnings that clutter output:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="s1">&#39;ignore&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Set the gray grid <a class="reference external" href="https://plotnine.readthedocs.io/en/stable/generated/plotnine.themes.theme.html?highlight=themes">plotnine theme</a> defined in <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.plotnine_themes.html">dms_variants.plotnine_themes</a> as this gives a nice appearance for the plots:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">theme_set</span><span class="p">(</span><span class="n">dms_variants</span><span class="o">.</span><span class="n">plotnine_themes</span><span class="o">.</span><span class="n">theme_graygrid</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="Simulate-library-of-variants">
<h2>Simulate library of variants<a class="headerlink" href="#Simulate-library-of-variants" title="Permalink to this headline">¶</a></h2>
<p>Simulate a library of codon variants of the gene.</p>
<p>Simulate wildtype gene sequence:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">geneseq</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">random</span><span class="o">.</span><span class="n">choices</span><span class="p">(</span><span class="n">CODONS_NOSTOP</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">genelength</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Wildtype gene of </span><span class="si">{genelength}</span><span class="s2"> codons:</span><span class="se">\n</span><span class="si">{geneseq}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Wildtype gene of 40 codons:
AGATCCGTGATTCTGCGTGCTTACACCAACTCACGGGTGAAACGTGTAATCTTATGCAACAACGACTTACCTATCCGCAACATCCGGCTGATGATGATCCTACACAACTCCGACGCTAGT
</pre></div></div>
</div>
<p>Generate a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a> using <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.simulate_CodonVariantTable">simulate_CodonVariantTable</a> function:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulate_CodonVariantTable</span><span class="p">(</span>
                <span class="n">geneseq</span><span class="o">=</span><span class="n">geneseq</span><span class="p">,</span>
                <span class="n">bclen</span><span class="o">=</span><span class="n">bclen</span><span class="p">,</span>
                <span class="n">library_specs</span><span class="o">=</span><span class="p">{</span><span class="n">lib</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;avgmuts&#39;</span><span class="p">:</span> <span class="n">avgmuts</span><span class="p">,</span>
                                     <span class="s1">&#39;nvariants&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span><span class="p">}</span>
                               <span class="k">for</span> <span class="n">lib</span> <span class="ow">in</span> <span class="n">libs</span><span class="p">},</span>
                <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
                <span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the number of amino-acid mutations per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_17_0.png" src="_images/narrow_bottleneck_17_0.png" />
</div>
</div>
<p>Average number of codon mutations per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_19_0.png" src="_images/narrow_bottleneck_19_0.png" />
</div>
</div>
<p>Numerical data in the plot above:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">numCodonMutsByType</span><span class="p">(</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>mutation_type</th>
      <th>num_muts_count</th>
      <th>count</th>
      <th>number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>nonsynonymous</td>
      <td>36281</td>
      <td>20000</td>
      <td>1.814</td>
    </tr>
    <tr>
      <td>1</td>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>synonymous</td>
      <td>1835</td>
      <td>20000</td>
      <td>0.092</td>
    </tr>
    <tr>
      <td>2</td>
      <td>lib_1</td>
      <td>barcoded variants</td>
      <td>stop</td>
      <td>1867</td>
      <td>20000</td>
      <td>0.093</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Examine how well amino-acid mutations are sampled in the library by looking at the fraction of mutations seen &lt;= some number of times, making separate plots for single mutants and all mutants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">variant_type</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;single&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulMutCoverage</span><span class="p">(</span><span class="n">variant_type</span><span class="p">,</span>
                                      <span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span>
                                      <span class="n">samples</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_23_0.png" src="_images/narrow_bottleneck_23_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_23_1.png" src="_images/narrow_bottleneck_23_1.png" />
</div>
</div>
</div>
<div class="section" id="Simulate-counts-for-samples">
<h2>Simulate counts for samples<a class="headerlink" href="#Simulate-counts-for-samples" title="Permalink to this headline">¶</a></h2>
<p>Now we simulate the counts of each variant after selection.</p>
<p>First, we define a “phenotype” function using a <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator">SigmoidPhenotypeSimulator</a>, which follow the “global epistasis” concepts of <a class="reference external" href="https://doi.org/10.1073/pnas.1804015115">Otwinoski et al</a> and <a class="reference external" href="https://doi.org/10.1534/genetics.116.195214">Sailer and Harms</a> to define the phenotype in two steps: an underlying latent phenotype that mutations affect additively, and then an
observed phenotype that is a non-linear function of the latent phenotype. The variants are then simulated according to their observed enrichments, which are the exponentials of the observed phenotypes:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">phenosimulator</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">SigmoidPhenotypeSimulator</span><span class="p">(</span>
                                            <span class="n">geneseq</span><span class="p">,</span>
                                            <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the simulated relationship of the latent phenotype with the observed enrichment and phenotype, with a dashed vertical line indicating the wildtype latent phenotype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;phenotype&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">phenosimulator</span><span class="o">.</span><span class="n">plotLatentVsObserved</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_27_0.png" src="_images/narrow_bottleneck_27_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_27_1.png" src="_images/narrow_bottleneck_27_1.png" />
</div>
</div>
<p>Plot the latent phenotype, observed phenotype, and observed enrichment of all single amino-acid mutants, with a dashed vertical line indicating the wildtype:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;latentPhenotype&#39;</span><span class="p">,</span> <span class="s1">&#39;observedPhenotype&#39;</span><span class="p">,</span> <span class="s1">&#39;observedEnrichment&#39;</span><span class="p">]:</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">phenosimulator</span><span class="o">.</span><span class="n">plotMutsHistogram</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
    <span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_29_0.png" src="_images/narrow_bottleneck_29_0.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_29_1.png" src="_images/narrow_bottleneck_29_1.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_29_2.png" src="_images/narrow_bottleneck_29_2.png" />
</div>
</div>
<p>Now we use <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.simulateSampleCounts">simulateSampleCounts</a> to simulate counts of variants when selection on each variant is proportional to its observed enrichment. We simulate using the several different bottleneck sizes defined above:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">counts</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">simulate</span><span class="o">.</span><span class="n">simulateSampleCounts</span><span class="p">(</span>
        <span class="n">variants</span><span class="o">=</span><span class="n">variants</span><span class="p">,</span>
        <span class="n">phenotype_func</span><span class="o">=</span><span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedEnrichment</span><span class="p">,</span>
        <span class="n">variant_error_rate</span><span class="o">=</span><span class="n">variant_error_rate</span><span class="p">,</span>
        <span class="n">pre_sample</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;total_count&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span><span class="n">avgdepth_per_variant</span><span class="p">,</span>
                    <span class="s1">&#39;uniformity&#39;</span><span class="p">:</span> <span class="n">lib_uniformity</span><span class="p">},</span>
        <span class="n">pre_sample_name</span><span class="o">=</span><span class="s1">&#39;pre-selection&#39;</span><span class="p">,</span>
        <span class="n">post_samples</span><span class="o">=</span><span class="p">{</span><span class="n">name</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;noise&#39;</span><span class="p">:</span> <span class="n">noise</span><span class="p">,</span>
                             <span class="s1">&#39;total_count&#39;</span><span class="p">:</span> <span class="n">variants_per_lib</span> <span class="o">*</span> <span class="n">avgdepth_per_variant</span><span class="p">,</span>
                             <span class="s1">&#39;bottleneck&#39;</span><span class="p">:</span> <span class="n">bottle</span><span class="p">}</span>
                          <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">bottle</span> <span class="ow">in</span> <span class="n">bottlenecks</span><span class="o">.</span><span class="n">items</span><span class="p">()},</span>
        <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">,</span>
        <span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now add the simulated counts for each library / sample to the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable">CodonVariantTable</a>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">add_sample_counts_df</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Confirm that we have added the expected number of counts per library / sample:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants</span><span class="o">.</span><span class="n">n_variants_df</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>count</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>2000000</td>
    </tr>
    <tr>
      <td>1</td>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>2000000</td>
    </tr>
    <tr>
      <td>2</td>
      <td>lib_1</td>
      <td>mid_bottle</td>
      <td>2000000</td>
    </tr>
    <tr>
      <td>3</td>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>2000000</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Plot the number of counts for each variant in each sample. The horizontal dashed line shows the total number of variants. The plot shows that all variants are well-sampled in the pre-selection libraries, but that post- selection some variants are sampled more or less. This is expected since selection will decrease and increase the frequency of variants:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotCumulVariantCounts</span><span class="p">(</span><span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;v&#39;</span><span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_37_0.png" src="_images/narrow_bottleneck_37_0.png" />
</div>
</div>
<p>Distribution of the number of amino-acid mutations per variant in each sample. As expected, mutations go down after selection:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumMutsHistogram</span><span class="p">(</span><span class="n">mut_type</span><span class="o">=</span><span class="s1">&#39;aa&#39;</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;v&#39;</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_39_0.png" src="_images/narrow_bottleneck_39_0.png" />
</div>
</div>
<p>Average number of mutations per variant:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">plotNumCodonMutsByType</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">orientation</span><span class="o">=</span><span class="s1">&#39;v&#39;</span><span class="p">)</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">p</span> <span class="o">+</span> <span class="n">theme</span><span class="p">(</span><span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">())</span>  <span class="c1"># no vertical grid lines</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_41_0.png" src="_images/narrow_bottleneck_41_0.png" />
</div>
</div>
<p>Here are the numerical data plotted above (just for the individual libraries and all mutations):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">variants</span><span class="o">.</span><span class="n">numCodonMutsByType</span><span class="p">(</span><span class="n">variant_type</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">,</span> <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span><span class="p">)</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>library</th>
      <th>sample</th>
      <th>mutation_type</th>
      <th>num_muts_count</th>
      <th>count</th>
      <th>number</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>0</td>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>nonsynonymous</td>
      <td>3617521</td>
      <td>2000000</td>
      <td>1.809</td>
    </tr>
    <tr>
      <td>1</td>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>synonymous</td>
      <td>184382</td>
      <td>2000000</td>
      <td>0.092</td>
    </tr>
    <tr>
      <td>2</td>
      <td>lib_1</td>
      <td>pre-selection</td>
      <td>stop</td>
      <td>187549</td>
      <td>2000000</td>
      <td>0.094</td>
    </tr>
    <tr>
      <td>3</td>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>nonsynonymous</td>
      <td>1824145</td>
      <td>2000000</td>
      <td>0.912</td>
    </tr>
    <tr>
      <td>4</td>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>synonymous</td>
      <td>174869</td>
      <td>2000000</td>
      <td>0.087</td>
    </tr>
    <tr>
      <td>5</td>
      <td>lib_1</td>
      <td>loose_bottle</td>
      <td>stop</td>
      <td>690</td>
      <td>2000000</td>
      <td>0.000</td>
    </tr>
    <tr>
      <td>6</td>
      <td>lib_1</td>
      <td>mid_bottle</td>
      <td>nonsynonymous</td>
      <td>1815987</td>
      <td>2000000</td>
      <td>0.908</td>
    </tr>
    <tr>
      <td>7</td>
      <td>lib_1</td>
      <td>mid_bottle</td>
      <td>synonymous</td>
      <td>172366</td>
      <td>2000000</td>
      <td>0.086</td>
    </tr>
    <tr>
      <td>8</td>
      <td>lib_1</td>
      <td>mid_bottle</td>
      <td>stop</td>
      <td>556</td>
      <td>2000000</td>
      <td>0.000</td>
    </tr>
    <tr>
      <td>9</td>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>nonsynonymous</td>
      <td>1825332</td>
      <td>2000000</td>
      <td>0.913</td>
    </tr>
    <tr>
      <td>10</td>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>synonymous</td>
      <td>166731</td>
      <td>2000000</td>
      <td>0.083</td>
    </tr>
    <tr>
      <td>11</td>
      <td>lib_1</td>
      <td>tight_bottle</td>
      <td>stop</td>
      <td>1016</td>
      <td>2000000</td>
      <td>0.001</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</div>
<div class="section" id="Functional-scores-for-variants">
<h2>Functional scores for variants<a class="headerlink" href="#Functional-scores-for-variants" title="Permalink to this headline">¶</a></h2>
<p>Use <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.func_scores">CodonVariantTable.func_scores</a> to calculates a functional score for each variant based on its change in frequency from pre- to post-selection:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">func_scores</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="s1">&#39;pre-selection&#39;</span><span class="p">,</span>
                                   <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Also calculate functional scores at the level of amino-acid variants:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">aa_func_scores</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">func_scores</span><span class="p">(</span><span class="s1">&#39;pre-selection&#39;</span><span class="p">,</span>
                                      <span class="n">by</span><span class="o">=</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span>
                                      <span class="n">syn_as_wt</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                      <span class="n">libraries</span><span class="o">=</span><span class="n">libs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Classify variants by the “types” of mutations they have using the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.codonvarianttable.html#dms_variants.codonvarianttable.CodonVariantTable.classifyVariants">CodonVariantTable.classifyVariants</a>:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">func_scores</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">codonvarianttable</span><span class="o">.</span><span class="n">CodonVariantTable</span><span class="o">.</span><span class="n">classifyVariants</span><span class="p">(</span><span class="n">func_scores</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Plot the distributions of scores, coloring by the variant class. As expected, there is more noise with a tighter bottleneck:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">ggplot</span><span class="p">(</span><span class="n">func_scores</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;variant_class&#39;</span><span class="p">,</span> <span class="s1">&#39;func_score&#39;</span><span class="p">))</span> <span class="o">+</span>
    <span class="n">geom_violin</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">fill</span><span class="o">=</span><span class="s1">&#39;variant_class&#39;</span><span class="p">))</span> <span class="o">+</span>
    <span class="n">ylab</span><span class="p">(</span><span class="s1">&#39;functional score&#39;</span><span class="p">)</span> <span class="o">+</span>
    <span class="n">xlab</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="o">+</span>
    <span class="n">facet_wrap</span><span class="p">(</span><span class="s1">&#39;~ post_sample&#39;</span><span class="p">)</span> <span class="o">+</span>
    <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">2.75</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">bottlenecks</span><span class="p">),</span>
                       <span class="mi">2</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">libs</span><span class="p">)),</span>
          <span class="n">axis_text_x</span><span class="o">=</span><span class="n">element_text</span><span class="p">(</span><span class="n">angle</span><span class="o">=</span><span class="mi">90</span><span class="p">),</span>
          <span class="n">panel_grid_major_x</span><span class="o">=</span><span class="n">element_blank</span><span class="p">(),</span>  <span class="c1"># no vertical grid lines</span>
          <span class="p">)</span> <span class="o">+</span>
    <span class="n">scale_fill_manual</span><span class="p">(</span><span class="n">values</span><span class="o">=</span><span class="n">CBPALETTE</span><span class="p">[</span><span class="mi">1</span> <span class="p">:],</span> <span class="n">guide</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_51_0.png" src="_images/narrow_bottleneck_51_0.png" />
</div>
</div>
</div>
<div class="section" id="Fit-global-epistasis-models">
<h2>Fit global epistasis models<a class="headerlink" href="#Fit-global-epistasis-models" title="Permalink to this headline">¶</a></h2>
<p>Now we fit global epistasis models. We fit two different types of models each with two different ways of calculating the likelihood.</p>
<p>The two types of models are: - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasis">MonotonicSplineEpistasis</a> models, which model global epistasis in the way described in <a class="reference external" href="https://www.pnas.org/content/115/32/E7550">Otwinoski et al (2018)</a>. - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.NoEpistasis">NoEpistasis</a> models, which assume mutations
contribute additively to the functional scores.</p>
<p>The two ways of calculating the likelihood are: - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.GaussianLikelihood">GaussianLikelihood</a>: assumes deviations from model are normally distributed according to statistical estimates of errors in functional scores. - <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.GaussianLikelihood">CauchyLikelihood</a>: a fat-tailed distribution that
is less sensitive to measurements that have large deviations from the model.</p>
<p>See the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/codonvariant_sim_data.html">globalepistasis module documentation</a> for more details:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">models</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># store models, keyed by `(modeltype, likelihoodtype, sample)`</span>

<span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">scores</span> <span class="ow">in</span> <span class="n">func_scores</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;post_sample&#39;</span><span class="p">,</span> <span class="n">observed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>

    <span class="n">bmap</span> <span class="o">=</span> <span class="n">dms_variants</span><span class="o">.</span><span class="n">binarymap</span><span class="o">.</span><span class="n">BinaryMap</span><span class="p">(</span><span class="n">scores</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">epistasistype</span><span class="p">,</span> <span class="n">liktype</span><span class="p">,</span> <span class="n">Model</span> <span class="ow">in</span> <span class="p">[</span>
            <span class="p">(</span><span class="s1">&#39;global epistasis&#39;</span><span class="p">,</span> <span class="s1">&#39;Gaussian&#39;</span><span class="p">,</span>
             <span class="n">dms_variants</span><span class="o">.</span><span class="n">globalepistasis</span><span class="o">.</span><span class="n">MonotonicSplineEpistasisGaussianLikelihood</span><span class="p">),</span>
            <span class="p">(</span><span class="s1">&#39;no epistasis&#39;</span><span class="p">,</span> <span class="s1">&#39;Gaussian&#39;</span><span class="p">,</span>
             <span class="n">dms_variants</span><span class="o">.</span><span class="n">globalepistasis</span><span class="o">.</span><span class="n">NoEpistasisGaussianLikelihood</span><span class="p">),</span>
            <span class="p">(</span><span class="s1">&#39;global epistasis&#39;</span><span class="p">,</span> <span class="s1">&#39;Cauchy&#39;</span><span class="p">,</span>
             <span class="n">dms_variants</span><span class="o">.</span><span class="n">globalepistasis</span><span class="o">.</span><span class="n">MonotonicSplineEpistasisCauchyLikelihood</span><span class="p">),</span>
            <span class="p">(</span><span class="s1">&#39;no epistasis&#39;</span><span class="p">,</span> <span class="s1">&#39;Cauchy&#39;</span><span class="p">,</span>
             <span class="n">dms_variants</span><span class="o">.</span><span class="n">globalepistasis</span><span class="o">.</span><span class="n">NoEpistasisCauchyLikelihood</span><span class="p">),</span>
            <span class="p">]:</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Fitting </span><span class="si">{epistasistype}</span><span class="s2"> model with </span><span class="si">{liktype}</span><span class="s2"> likelihood to </span><span class="si">{sample}</span><span class="s2">...&quot;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39; &#39;</span><span class="p">)</span>

        <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">Model</span><span class="p">(</span><span class="n">bmap</span><span class="p">)</span>
        <span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;fitting took {time.time() - start:.1f} sec.&quot;</span><span class="p">)</span>
        <span class="n">models</span><span class="p">[(</span><span class="n">epistasistype</span><span class="p">,</span> <span class="n">liktype</span><span class="p">,</span> <span class="n">sample</span><span class="p">)]</span> <span class="o">=</span> <span class="n">model</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Fitting global epistasis model with Gaussian likelihood to loose_bottle... fitting took 44.4 sec.
Fitting no epistasis model with Gaussian likelihood to loose_bottle... fitting took 0.3 sec.
Fitting global epistasis model with Cauchy likelihood to loose_bottle... fitting took 42.3 sec.
Fitting no epistasis model with Cauchy likelihood to loose_bottle... fitting took 2.2 sec.
Fitting global epistasis model with Gaussian likelihood to mid_bottle... fitting took 16.4 sec.
Fitting no epistasis model with Gaussian likelihood to mid_bottle... fitting took 0.2 sec.
Fitting global epistasis model with Cauchy likelihood to mid_bottle... fitting took 22.7 sec.
Fitting no epistasis model with Cauchy likelihood to mid_bottle... fitting took 1.3 sec.
Fitting global epistasis model with Gaussian likelihood to tight_bottle... fitting took 12.3 sec.
Fitting no epistasis model with Gaussian likelihood to tight_bottle... fitting took 0.2 sec.
Fitting global epistasis model with Cauchy likelihood to tight_bottle... fitting took 21.8 sec.
Fitting no epistasis model with Cauchy likelihood to tight_bottle... fitting took 1.3 sec.
</pre></div></div>
</div>
<p>Now we want to see which models fit the data better. To do this, we get the log likelihood of each model along with the number of model parameters and use it to calculate the <a class="reference external" href="https://en.wikipedia.org/wiki/Akaike_information_criterion">AIC</a>. Models with lower AIC are better. We compute the <span class="math notranslate nohighlight">\(\Delta\)</span>AIC between the global epistasis model
(<a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasis">MonotonicSplineEpistasis</a>) and no-epistasis model (<a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.NoEpistasis">NoEpistasis</a>): positive <span class="math notranslate nohighlight">\(\Delta\)</span>AIC values indicate that the global epistasis model is better.</p>
<p>Crucially, we can only compare models with likelihoods calculated in the same way: Gaussian likelihoods can be compared to each other, Cauchy likelilhoods can be compared to each other, but Gaussian and Cauchy likelihoods can <strong>not</strong> be compared.</p>
<p>Below we see that in all cases, the global epistasis model is better than the no-epistasis model:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">logliks_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_records</span><span class="p">(</span>
            <span class="p">[(</span><span class="n">epistasistype</span><span class="p">,</span> <span class="n">liktype</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">nparams</span><span class="p">,</span> <span class="n">model</span><span class="o">.</span><span class="n">loglik</span><span class="p">)</span> <span class="k">for</span>
             <span class="p">(</span><span class="n">epistasistype</span><span class="p">,</span> <span class="n">liktype</span><span class="p">,</span> <span class="n">sample</span><span class="p">),</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">()],</span>
            <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;epistasis_function&#39;</span><span class="p">,</span> <span class="s1">&#39;likelihood&#39;</span><span class="p">,</span> <span class="s1">&#39;sample&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;n_parameters&#39;</span><span class="p">,</span> <span class="s1">&#39;log_likelihood&#39;</span><span class="p">]</span>
            <span class="p">)</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">AIC</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;n_parameters&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;log_likelihood&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;likelihood&#39;</span><span class="p">],</span>
                 <span class="n">values</span><span class="o">=</span><span class="s1">&#39;AIC&#39;</span><span class="p">,</span>
                 <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;epistasis_function&#39;</span><span class="p">,</span>
                 <span class="p">)</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">deltaAIC</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;no epistasis&#39;</span><span class="p">]</span> <span class="o">-</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;global epistasis&#39;</span><span class="p">])</span>
    <span class="p">)</span>

<span class="n">logliks_df</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>epistasis_function</th>
      <th>global epistasis</th>
      <th>no epistasis</th>
      <th>deltaAIC</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>likelihood</th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td rowspan="2" valign="top">loose_bottle</td>
      <td>Cauchy</td>
      <td>46817.6</td>
      <td>68553.9</td>
      <td>21736.3</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>36921.0</td>
      <td>79161.8</td>
      <td>42240.8</td>
    </tr>
    <tr>
      <td rowspan="2" valign="top">mid_bottle</td>
      <td>Cauchy</td>
      <td>73421.0</td>
      <td>91584.6</td>
      <td>18163.6</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>71744.1</td>
      <td>91161.0</td>
      <td>19416.9</td>
    </tr>
    <tr>
      <td rowspan="2" valign="top">tight_bottle</td>
      <td>Cauchy</td>
      <td>95439.8</td>
      <td>113039.5</td>
      <td>17599.7</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>102515.5</td>
      <td>106334.1</td>
      <td>3818.6</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Because the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.MonotonicSplineEpistasis">MonotonicSplineEpistasis</a> global epistasis model is always better than the additive <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.NoEpistasis">NoEpistasis</a> model, below we just analyze the results from the global epistasis model.</p>
<p>First, we will examine how the model fits look on all the actual variants used to fit the model. We use <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.phenotypes_df">AbstractEpistasis.phenotypes_df</a> to get a data frame of all the variants used to fit each global epistasis model along with their functional scores and the latent and observed phenotypes predicted by each model:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span><span class="n">model</span><span class="o">.</span><span class="n">phenotypes_df</span>
         <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">sample</span><span class="o">=</span><span class="n">sample</span><span class="p">,</span>
                 <span class="n">likelihood</span><span class="o">=</span><span class="n">liktype</span><span class="p">,</span>
                 <span class="p">)</span>
         <span class="k">for</span> <span class="p">(</span><span class="n">epistasistype</span><span class="p">,</span> <span class="n">liktype</span><span class="p">,</span> <span class="n">sample</span><span class="p">),</span> <span class="n">model</span> <span class="ow">in</span> <span class="n">models</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
         <span class="k">if</span> <span class="n">epistasistype</span> <span class="o">==</span> <span class="s1">&#39;global epistasis&#39;</span><span class="p">],</span>
        <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Below we plot the relationships between the observed and latent phenotype for each sample and likelihood-calculation method:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">variants_df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;latent_phenotype&#39;</span><span class="p">,</span> <span class="s1">&#39;observed_phenotype&#39;</span><span class="p">))</span> <span class="o">+</span>
     <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;sample ~ likelihood&#39;</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span><span class="s1">&#39;free&#39;</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;likelihood&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
                        <span class="mi">2</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">()),</span>
           <span class="p">)</span>
     <span class="p">)</span>
<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_60_0.png" src="_images/narrow_bottleneck_60_0.png" />
</div>
</div>
</div>
<div class="section" id="Model-vs-measurements-vs-truth-for-amino-acid-variants">
<h2>Model vs measurements vs truth for amino-acid variants<a class="headerlink" href="#Model-vs-measurements-vs-truth-for-amino-acid-variants" title="Permalink to this headline">¶</a></h2>
<p>Because we simulated the variants, we can also use the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.simulate.html#dms_variants.simulate.SigmoidPhenotypeSimulator">SigmoidPhenotypeSimulator</a> used in these simulations to get the <strong>true</strong> phenotype of each variant. (Obviously in real non-simulated experiments the true phenotypes are unknown.)</p>
<p>We want to see if the model outperforms the experiments if look at the level of <strong>amino-acid</strong> variants (meaning we pool the counts for all variants with the same amino-acid substitutions together before calculating the functional scores). Recall that above we already got such functional scores into the variable <code class="docutils literal notranslate"><span class="pre">aa_func_scores</span></code>:</p>
<p>First we get a data frame of measured functional scores (in the simulated experiment) calculated at the level of amino-acid variants (pooling all barcoded variants with the same amino-acid mutations):</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">aa_variants_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_func_scores</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;post_sample&#39;</span><span class="p">:</span> <span class="s1">&#39;sample&#39;</span><span class="p">})</span>
    <span class="p">[[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;n_aa_substitutions&#39;</span><span class="p">,</span> <span class="s1">&#39;func_score&#39;</span><span class="p">]]</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">measured_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="mi">2</span><span class="o">**</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;func_score&#39;</span><span class="p">])</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now we add the model-predicted latent phenotype, observed phenotype, and enrichment for each amino-acid variant using the <a class="reference external" href="https://jbloomlab.github.io/dms_variants/dms_variants.globalepistasis.html#dms_variants.globalepistasis.AbstractEpistasis.add_phenotypes_to_df">AbstractEpistasis.add_phenotypes_to_df</a> method:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">df_list</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">sample</span><span class="p">,</span> <span class="n">df</span> <span class="ow">in</span> <span class="n">aa_variants_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="n">observed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">likelihood</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;Cauchy&#39;</span><span class="p">,</span> <span class="s1">&#39;Gaussian&#39;</span><span class="p">]:</span>
        <span class="n">model</span> <span class="o">=</span> <span class="n">models</span><span class="p">[(</span><span class="s1">&#39;global epistasis&#39;</span><span class="p">,</span> <span class="n">likelihood</span><span class="p">,</span> <span class="n">sample</span><span class="p">)]</span>
        <span class="n">df_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">model</span><span class="o">.</span><span class="n">add_phenotypes_to_df</span><span class="p">(</span><span class="n">df</span><span class="p">)</span>
                       <span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">predicted_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span>
                                  <span class="n">model</span><span class="o">.</span><span class="n">enrichments</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;observed_phenotype&#39;</span><span class="p">]),</span>
                               <span class="n">likelihood</span><span class="o">=</span><span class="n">likelihood</span><span class="p">,</span>
                           <span class="p">)</span>
                   <span class="p">)</span>
<span class="n">aa_variants_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span><span class="n">df_list</span><span class="p">,</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now we add the “true” values from the simulator:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">aa_variants_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_variants_df</span>
    <span class="o">.</span><span class="n">assign</span><span class="p">(</span>
        <span class="n">true_latent_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">latentPhenotype</span><span class="p">),</span>
        <span class="n">true_enrichment</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedEnrichment</span><span class="p">),</span>
        <span class="n">true_observed_phenotype</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;aa_substitutions&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span>
                <span class="n">phenosimulator</span><span class="o">.</span><span class="n">observedPhenotype</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<p>Now we calculate the correlations between the true enrichments from the simulations and the measured values and the ones predicted by the model. We do this using enrichment rather than observed phenotype as we expect the observed phenotypes to be very noisy at the “low end” (since experiments lose the sensitivity to distinguish among bad and really-bad variants):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">enrichments_corr</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_variants_df</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;model prediction&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;measured_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;measurement&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;likelihood&#39;</span><span class="p">,</span> <span class="s1">&#39;true_enrichment&#39;</span><span class="p">],</span>
          <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;model prediction&#39;</span><span class="p">,</span> <span class="s1">&#39;measurement&#39;</span><span class="p">],</span>
          <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">,</span>
          <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span>
          <span class="p">)</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;likelihood&#39;</span><span class="p">,</span> <span class="s1">&#39;enrichment_type&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;enrichment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">]))</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;correlation&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;likelihood&#39;</span><span class="p">],</span>
                 <span class="n">values</span><span class="o">=</span><span class="s1">&#39;correlation&#39;</span><span class="p">,</span>
                 <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">)</span>
    <span class="p">)</span>

<span class="n">enrichments_corr</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>enrichment_type</th>
      <th>measurement</th>
      <th>model prediction</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>likelihood</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td rowspan="2" valign="top">loose_bottle</td>
      <td>Cauchy</td>
      <td>0.98</td>
      <td>0.99</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>0.98</td>
      <td>1.00</td>
    </tr>
    <tr>
      <td rowspan="2" valign="top">mid_bottle</td>
      <td>Cauchy</td>
      <td>0.82</td>
      <td>0.99</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>0.82</td>
      <td>0.99</td>
    </tr>
    <tr>
      <td rowspan="2" valign="top">tight_bottle</td>
      <td>Cauchy</td>
      <td>0.64</td>
      <td>0.93</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>0.64</td>
      <td>0.79</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Above we see that the model predictions are always better than the actual measurements, and furthermore that when there is a very tight experimental bottleneck (lots of “noise”) the Cauchy likelihood outperforms the Gaussian likelihood.</p>
<p>Below we plot the correlations between the model predictions and the true enrichments for each likelihood calculation method. The plot shows that the Cauchy likelihood seems better than the Gaussian likelihood for the tight bottleneck, although it appears visually that for a loose bottleneck (relatively little experimental noise) the Guassian likelihood might be better:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[37]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">aa_variants_df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">))</span> <span class="o">+</span>
     <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;sample ~ likelihood&#39;</span><span class="p">,</span>  <span class="n">scales</span><span class="o">=</span><span class="s1">&#39;free&#39;</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">2.5</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;likelihood&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
                        <span class="mf">2.5</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">())</span>
           <span class="p">)</span>
     <span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_71_0.png" src="_images/narrow_bottleneck_71_0.png" />
</div>
</div>
</div>
<div class="section" id="Model-vs-measurements-vs-truth-for-single-mutants">
<h2>Model vs measurements vs truth for single mutants<a class="headerlink" href="#Model-vs-measurements-vs-truth-for-single-mutants" title="Permalink to this headline">¶</a></h2>
<p>Now we repeat the above analysis for just <strong>single</strong> amino-acid mutants.</p>
<p>First, get the single amino-acid mutant variants:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">single_aa_variants_df</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">aa_variants_df</span>
    <span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="s1">&#39;n_aa_substitutions == 1&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="p">)</span>
</pre></div>
</div>
</div>
<p>Tabulate how the measurements and model predictions compare to the true enrichments:</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="c1"># NBVAL_IGNORE_OUTPUT</span>

<span class="n">enrichments_corr</span> <span class="o">=</span> <span class="p">(</span>
    <span class="n">single_aa_variants_df</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;model prediction&#39;</span><span class="p">,</span>
                     <span class="s1">&#39;measured_enrichment&#39;</span><span class="p">:</span> <span class="s1">&#39;measurement&#39;</span><span class="p">})</span>
    <span class="o">.</span><span class="n">melt</span><span class="p">(</span><span class="n">id_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;likelihood&#39;</span><span class="p">,</span> <span class="s1">&#39;true_enrichment&#39;</span><span class="p">],</span>
          <span class="n">value_vars</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;model prediction&#39;</span><span class="p">,</span> <span class="s1">&#39;measurement&#39;</span><span class="p">],</span>
          <span class="n">var_name</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">,</span>
          <span class="n">value_name</span><span class="o">=</span><span class="s1">&#39;enrichment&#39;</span><span class="p">,</span>
          <span class="p">)</span>
    <span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;likelihood&#39;</span><span class="p">,</span> <span class="s1">&#39;enrichment_type&#39;</span><span class="p">])</span>
    <span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="p">[</span><span class="s1">&#39;enrichment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">]))</span>
    <span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="s1">&#39;correlation&#39;</span><span class="p">)</span>
    <span class="o">.</span><span class="n">to_frame</span><span class="p">()</span>
    <span class="o">.</span><span class="n">pivot_table</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">,</span> <span class="s1">&#39;likelihood&#39;</span><span class="p">],</span>
                 <span class="n">values</span><span class="o">=</span><span class="s1">&#39;correlation&#39;</span><span class="p">,</span>
                 <span class="n">columns</span><span class="o">=</span><span class="s1">&#39;enrichment_type&#39;</span><span class="p">)</span>
    <span class="p">)</span>

<span class="n">enrichments_corr</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>enrichment_type</th>
      <th>measurement</th>
      <th>model prediction</th>
    </tr>
    <tr>
      <th>sample</th>
      <th>likelihood</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td rowspan="2" valign="top">loose_bottle</td>
      <td>Cauchy</td>
      <td>0.99</td>
      <td>1.00</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>0.99</td>
      <td>1.00</td>
    </tr>
    <tr>
      <td rowspan="2" valign="top">mid_bottle</td>
      <td>Cauchy</td>
      <td>0.90</td>
      <td>1.00</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>0.90</td>
      <td>1.00</td>
    </tr>
    <tr>
      <td rowspan="2" valign="top">tight_bottle</td>
      <td>Cauchy</td>
      <td>0.77</td>
      <td>0.93</td>
    </tr>
    <tr>
      <td>Gaussian</td>
      <td>0.77</td>
      <td>0.77</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Once again, the model predictions outperform the measurements, and the Cauchy likelihood is noticeably better than the Gaussian likelihood for the tight bottleneck.</p>
<p>Plot the predictions for each likelihood-calculation method. The plots below visually make clear that the Cauchy likelihood is better for the tight bottleneck (lots of noise). It again visually appears that the Gaussian likelihood is a bit better for the loose bottleneck (very little noise):</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">p</span> <span class="o">=</span> <span class="p">(</span><span class="n">ggplot</span><span class="p">(</span><span class="n">single_aa_variants_df</span><span class="p">,</span> <span class="n">aes</span><span class="p">(</span><span class="s1">&#39;true_enrichment&#39;</span><span class="p">,</span> <span class="s1">&#39;predicted_enrichment&#39;</span><span class="p">))</span> <span class="o">+</span>
     <span class="n">geom_point</span><span class="p">(</span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">facet_grid</span><span class="p">(</span><span class="s1">&#39;sample ~ likelihood&#39;</span><span class="p">,</span>  <span class="n">scales</span><span class="o">=</span><span class="s1">&#39;free&#39;</span><span class="p">)</span> <span class="o">+</span>
     <span class="n">theme</span><span class="p">(</span><span class="n">figure_size</span><span class="o">=</span><span class="p">(</span><span class="mf">2.5</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;likelihood&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">(),</span>
                        <span class="mf">2.5</span> <span class="o">*</span> <span class="n">variants_df</span><span class="p">[</span><span class="s1">&#39;sample&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">nunique</span><span class="p">())</span>
           <span class="p">)</span>
     <span class="p">)</span>

<span class="n">_</span> <span class="o">=</span> <span class="n">p</span><span class="o">.</span><span class="n">draw</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/narrow_bottleneck_78_0.png" src="_images/narrow_bottleneck_78_0.png" />
</div>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">dms_variants</h1>
    
  </a>
</p>



<p class="blurb">Analyze deep mutational scanning of barcoded variants</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=dms_variants&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/dms_variants">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/dms_variants.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">dms_variants documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="codonvariant_sim_data.html">Analyze simulated library of codon variants</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Fitting gobal epistasis models with experimental “noise”</a></li>
<li class="toctree-l2"><a class="reference internal" href="predict_variants.html">Predict unseen variants using global epistasis models</a></li>
<li class="toctree-l2"><a class="reference internal" href="codonvariant_plot_formatting.html">Formatting <code class="docutils literal notranslate"><span class="pre">CodonVariantTable</span></code> plots</a></li>
<li class="toctree-l2"><a class="reference internal" href="parsebarcodes_sim_data.html">Parse barcodes</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="dms_variants.html">dms_variants package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="examples.html">Examples</a><ul>
      <li>Previous: <a href="codonvariant_sim_data.html" title="previous chapter">Analyze simulated library of codon variants</a></li>
      <li>Next: <a href="predict_variants.html" title="next chapter">Predict unseen variants using global epistasis models</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2019.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/narrow_bottleneck.nblink.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/dms_variants" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>